satpy/readers/utils.py

```diff
@@ -210,57 +233,84 @@ def unzip_file(filename, prefix=None):
-                raise IOError("pbzip2 error '%s', failed, status=%d"
-                              % (filename, status))
-            with closing(os.fdopen(fdn, 'wb')) as ofpt:
-                try:
-                    stdout.seek(0)
-                    shutil.copyfileobj(stdout, ofpt)
-                except IOError:
-                    import traceback
-                    traceback.print_exc()
-                    LOGGER.info("Failed to read bzipped file %s",
-                                str(filename))
-                    os.remove(tmpfilepath)
-                    raise
-            return tmpfilepath
-
-        # Otherwise, fall back to the original method
-        bz2file = bz2.BZ2File(filename)
-        with closing(os.fdopen(fdn, 'wb')) as ofpt:
-            try:
-                ofpt.write(bz2file.read())
-            except IOError:
-                import traceback
-                traceback.print_exc()
-                LOGGER.info("Failed to read bzipped file %s", str(filename))
-                os.remove(tmpfilepath)
-                return None
-        return tmpfilepath
-
-    return None
+    fdn, tmpfilepath = tempfile.mkstemp(prefix=prefix,
+                                        dir=config["tmp_dir"])
+    LOGGER.info("Using temp file for BZ2 decompression: %s", tmpfilepath)
+    # check pbzip2 status
+    pbzip2 = _unzip_with_pbzip(filename, tmpfilepath, fdn)
+    if pbzip2 is not None:
+        return pbzip2
+    # Otherwise, fall back to the original method
+    bz2file = bz2.BZ2File(filename)
+    content = bz2file.read()
+    return _write_uncompress_file(content, fdn, filename, tmpfilepath)
+
+
+def _unzip_with_pbzip(filename, tmpfilepath, fdn):
+    # try pbzip2
+    pbzip = which('pbzip2')
+    if pbzip is None:
+        return None
+    # Run external pbzip2
+    n_thr = os.environ.get('OMP_NUM_THREADS')
+    if n_thr:
+        runner = [pbzip,
+                  '-dc',
+                  '-p'+str(n_thr),
+                  filename]
+    else:
+        runner = [pbzip,
+                  '-dc',
+                  filename]
+    p = Popen(runner, stdout=PIPE, stderr=PIPE)  # nosec
+    stdout = BytesIO(p.communicate()[0])
+    status = p.returncode
+    if status != 0:
+        raise IOError("pbzip2 error '%s', failed, status=%d"
+                      % (filename, status))
+    with closing(os.fdopen(fdn, 'wb')) as ofpt:
+        try:
+            stdout.seek(0)
+            shutil.copyfileobj(stdout, ofpt)
+        except IOError:
+            LOGGER.debug("Failed to read bzipped file %s", str(filename))
+            os.remove(tmpfilepath)
+            raise
+    return tmpfilepath
+
+
+def _write_uncompress_file(content, fdn, filename, tmpfilepath):
+    with closing(os.fdopen(fdn, 'wb')) as ofpt:
+        try:
+            ofpt.write(content)
+        except IOError:
+            LOGGER.debug("Failed to read bzipped file %s", str(filename))
+            os.remove(tmpfilepath)
+            return None
```

Comment:
This should be place back around the `.read` call, as I think the IOError we want to catch if for reading, not writing.

Reply:
Maybe adding IOError catch for both reading and writing will be better?

Reply:
Maybe adding IOError catch for both reading and writing will be better?

Reply:
Yes, that works for me :)

Reply:
OK, leaving them unhandled is also an appropriate solution :)